### PR TITLE
Check spawned worker version vs node version before PVF preparation

### LINKED ISSUE #6860: Check worker version against node version before running execution/preparation jobs
As we learned during a recent incident, upgrading node binary in place without restart may lead to weird behavior and produce disputes. We need to properly report that kind of error and force the node to shut down in case of such misbehavior.

### PR DESCRIPTION
**Pull request by s0me0ne-unkn0wn**
Closes #6860 

Compare worker and node versions and force the node to shut down in case of a mismatch.

### CODE REVIEW

**Comment by s0me0ne-unkn0wn** [comment]
This is for pre-review right now. It the approach is okay, I'll also implement the same logic for execution workers.

File: cli/src/cli.rs
Comment (bkchr) [inline cli/src/cli.rs:84]: We should make this required and not optional.

File: node/core/pvf/src/prepare/worker.rs
Comment (bkchr) [inline node/core/pvf/src/prepare/worker.rs:58]:
```suggestion
		&["prepare-worker", "--node-impl-version", env!("SUBSTRATE_CLI_IMPL_VERSION")],
```

File: node/core/pvf/src/prepare/worker.rs
Comment (bkchr) [inline node/core/pvf/src/prepare/worker.rs:366]: Why isn't that done outside of the loop?

File: node/core/pvf/src/prepare/worker.rs
Comment (bkchr) [inline node/core/pvf/src/prepare/worker.rs:191]: I think you can send a signal to the overseer to let it shutdown?

**Review by bkchr — COMMENTED** [review]
Goes into the right direction!

File: node/core/pvf/src/prepare/worker.rs
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/prepare/worker.rs:366]: Node<->worker exchange is half-duplex, to send an error response, we must accept a request first. But yes, it's invariant to the loop body, so I should have placed that outside the loop and only fired it once after the first request. I'll fix it.

File: node/core/pvf/src/prepare/worker.rs
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/prepare/worker.rs:191]: It's PVF host, it isn't a subsystem by itself, and it's not aware of overseer's existence :roll_eyes: 
To handle it the right way, the error should be propagated all the way up to the candidate validation and PVF pre-check subsystems and handled by them separately. Do you think it's worth the effort?

Comment (mrcnski) [reply node/core/pvf/src/prepare/worker.rs:191]: You shouldn't just kill the process here, you should bubble up this error and let the host kill the worker properly. Look at how the host handles e.g. `TimedOut` for an example. Also, I'm pretty sure this code is running on the host so `exit` would kill the node, not the worker, which I think is not what we want? (Haven't followed the disputes issue that closely over the weekend.)

File: node/core/pvf/src/prepare/worker.rs
Comment (mrcnski) [inline node/core/pvf/src/prepare/worker.rs:362]: nit compliment: Good call on moving this out of the `loop`. 👍

**Review by mrcnski — CHANGES_REQUESTED** [review]
Interesting solution. Just needs a fix to kill the worker properly. I still have to catch up on the disputes issue and will give this PR a closer look then.

File: node/core/pvf/src/prepare/worker.rs
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/prepare/worker.rs:191]: > Also, I'm pretty sure this code is running on the host so `exit` would kill the node, not the worker

Yes, it's exactly what we want. At this point, we found out that the node owner screwed up the node upgrade process and we're still running an old node software, but the binary is already new, and we're trying to execute a new worker from the old node. In that case, we want to tear the whole node down and let the node owner handle its restart.

Comment (mrcnski) [reply node/core/pvf/src/prepare/worker.rs:191]: Oh, got it. Is any clean-up of the workers necessary, or do they get killed automatically as child processes?

Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/prepare/worker.rs:191]: Children are not killed automatically, but if the worker recognizes `version_mismatch`, it will `return` from the main loop which will terminate the process.

Other workers may still be alive at the moment, but after they get some outcome from their jobs they'll try to send it to the node through the unix socket that is already closed, which will also force them to terminate.

We can terminate them manually (dropping the workers' handles) before tearing down the node, but why? They'll exit anyway.

File: cli/src/command.rs
Comment (dsl2022) [inline cli/src/command.rs:500]: t

File: node/core/pvf/src/prepare/worker.rs
Comment (mrcnski) [reply node/core/pvf/src/prepare/worker.rs:191]: Please document this all clearly above the `exit`. There are already a number of silent assumptions being made throughout the pvf code, and we should at least have comments to make things more explicit. It is also good just to draw attention to the `exit` because it is quite an extreme measure to suddenly shut down the entire node.

File: node/core/pvf/src/prepare/worker.rs
Comment (mrcnski) [inline node/core/pvf/src/prepare/worker.rs:364]: So there are two hypothetical cases that can trigger this:

1. Old node, new worker -> the old node should be shutdown
1. New node, old worker -> the worker should be shutdown (and probably any result discarded and the job re-tried)

Can 2 ever happen? If so, we should not shutdown the whole node in that case.

Comment (bkchr) [reply node/core/pvf/src/prepare/worker.rs:364]: 2 can not happen as the new node doesn't know the old workers.

File: node/core/pvf/src/prepare/worker.rs
Comment (bkchr) [reply node/core/pvf/src/prepare/worker.rs:191]: We need to wait for the workers to shutdown. Otherwise we may put stuff into the cache that is invalid. Let's assume the following. Node at version X is spawning a preparation worker. Node at version X is restarted and starts again as node at version x + 1. The worker finishes and writes the wasm artifact into the cache. Then the cache contains an artifact from version X.

File: node/core/pvf/src/prepare/worker.rs
Comment (mrcnski) [reply node/core/pvf/src/prepare/worker.rs:364]: Yes, good point. This should be documented.

File: node/core/pvf/src/prepare/worker.rs
Comment (mrcnski) [reply node/core/pvf/src/prepare/worker.rs:191]: Good point. Unfortunately this means potentially waiting the full lenient timeout duration for a worker to exit. I think it would make sense to just kill them forcefully so we don't wait.

Alternatively, we could detect which artifacts were from an old version. This is useful anyways if we ever stop deleting the cache on restart (there is an ongoing discussion about this). But if we knew that artifacts were from an old worker and invalid, then we wouldn't need to shutdown workers.

File: node/core/pvf/src/prepare/worker.rs
Comment (ordian) [reply node/core/pvf/src/prepare/worker.rs:364]: > Can 2 ever happen?

No, I don't how that's possible.

File: node/core/pvf/src/prepare/worker.rs
Comment (ordian) [reply node/core/pvf/src/prepare/worker.rs:191]: It makes sense to implement @koute suggestion as well and include polkadot's version into the artifact path. This way old version can't override an artifact for the new node.

File: node/core/pvf/src/prepare/worker.rs
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/prepare/worker.rs:364]: Well, purely theoretically, one can downgrade the node in-place, then we'll get 2 :) But it's nothing different from the node upgrade scenario, we'll have to shutdown the node anyway. So I don't try to check if the worker version is greater or less, just "does not match" is enough.

File: node/core/pvf/src/prepare/worker.rs
Comment (bkchr) [reply node/core/pvf/src/prepare/worker.rs:191]: > Good point. Unfortunately this means potentially waiting the full lenient timeout duration for a worker to exit. I think it would make sense to just kill them forcefully so we don't wait.

Yes for sure we should kill the workers, they don't do anything important that could get lost when we kill them.

> It makes sense to implement @koute suggestion as well and include polkadot's version into the artifact path. This way old version can't override an artifact for the new node.

This should be a separate pr.

Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/prepare/worker.rs:191]: > We need to wait for the workers to shutdown.

Good point! But I've reviewed the code and looks like @pepyakin thought this through before us.

`start_work()` on the host side provides the worker with a temporary path where it should store the artifact. If the worker reports a positive result, then in `handle_result()`, again on the host side, the artifact gets renamed to its real `artifact_id` path. If the host dies between `start_work()` and `handle_result()`, the only leftover is a stale tmp file which will be pruned on the next node restart.

It's still a good idea to kill preparation workers on version mismatch.

Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/prepare/worker.rs:191]: @koute you mean some random number we generate on node startup and use throughout the node instance lifetime? I think it's a good approach also, but your first idea about including the node version number to the artifact name is somewhat more... Deterministic, I'd say? :)

**Comment by s0me0ne-unkn0wn** [comment]
Reworked it quite a bit. As @bkchr rightly observed, workers should be killed before the node shutdown. It's crucial to kill preparation workers, as execution workers do not leave anything behind. But the execution is a much more frequent event than the preparation, so an execution worker is much more likely to recognize a version mismatch first. Because of that, the only option to handle it properly is to propagate the error from any worker to the PVF host and let it command both pipelines to shut down. Then, after both of them reported back they killed all the workers, stopped all the activity, and are ready to be shut down, the PVF host tears the node down.

In principle, it's very close now to being handled the totally right way, that is, propagating the error one step higher to the candidate validation subsystem to signal the shutdown event to the overseer.

It's not complete yet, I definitely need some tests and logical checks.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2517528

File: node/core/pvf/src/execute/worker.rs
Comment (mrcnski) [inline node/core/pvf/src/execute/worker.rs:285]: Should document why we check this in the `loop` instead of where `version_mismatch` is set. It's the correct behavior but it's not immediately obvious why.

**Review by mrcnski — CHANGES_REQUESTED** [review]
Really good work on short notice. 👍 But... I'm not really comfortable with the significant increase in the complexity of the pipelines, to handle a quite rare corner case. It makes it harder to reason about and review and make future changes. 😬

For such an exceptional case I think we should enact an exceptional measure: just send a signal to all the child processes to force-kill them. Right at the spot where we first receive a `VersionMismatch` signal, list all forked processes and send `SIGKILL` or whatever, and don't bubble further up. Would that possible? Then we don't need to do all this coordination between the prepare and execute pipelines. It should be safe to kill the processes because they store artifacts in temporary locations.

File: node/core/pvf/src/execute/worker.rs
Comment (mrcnski) [reply node/core/pvf/src/execute/worker.rs:285]: Think we need to do another version check, between writing the temporary artifact and sending a success result back to the node.

Comment (mrcnski) [reply node/core/pvf/src/execute/worker.rs:285]: Should we add another version check on the node side, when receiving the temporary artifact, that it was built with the current node version, before renaming the artifact?

File: node/core/pvf/src/prepare/worker.rs
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/prepare/worker.rs:61]: Not only this one :) I'm testing right now and have already found some missing logic and bugs

Comment (mrcnski) [reply node/core/pvf/src/prepare/worker.rs:61]: Honestly I would say that's an argument in favor of simplifying this. With these pipelines it's very easy to miss subtle logic errors. 😬 Sorry.

File: node/core/pvf/src/execute/worker.rs
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/execute/worker.rs:285]: > Should we add another version check on the node side

I don't think so, the new node instance wouldn't be able to communicate with an old worker as they communicate through unix socket created for every worker from scratch.

**Comment by s0me0ne-unkn0wn** [comment]
> Right at the spot where we first receive a `VersionMismatch` signal, list all forked processes and send `SIGKILL` or whatever

I believe there's no portable way to list all the child processes (beyond calling `popen("ps ax --ppid")` and then parsing the output but that looks way too hacky even for corner-case handling). So the usual practice is to save their pids on `fork()` but in our case `fork()`s are abstracted by `tokio::process` handles and bookkeeping of those handles is done by preparation and execution queues separately. So I don't see a good option beyond bubbling errors up, it's annoying but very idiomatic approach :)

Let me know if you have better ideas, I'd also like to simplify this.

**Comment by mrcnski** [comment]
@s0me0ne-unkn0wn I believe we can just send `kill` with a pid of `-parent_pid` to kill both the node and all child processes. This works by killing all processes in the "process group" indexed by `parent_id`.

https://www.man7.org/linux/man-pages/man1/kill.1.html

This looks to be POSIX and cross-platform:

https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/kill.2.html

And I confirmed with Activity Monitor on my Mac that process groups are a thing here:

Screenshot 2023-03-14 at 19 32 31

I'm also now on the FBI watch list after searching "man kill".

**Comment by s0me0ne-unkn0wn** [comment]
> I believe we can just send `kill` with a pid of `-parent_pid` to kill both the node and all child processes.

Yeah, that's interesting, thank you! Also a convenient low-level interface is the `kill` syscall:

> If `pid` equals 0, then `sig` is sent to every process in the process group of the calling process.

and we already have `libc` as a PVF host dependency. I'll give it a try. My only concern is that the node will receive `SIGKILL` too, and what if some thread is writing to the database at the moment? Not sure how critical it is. @bkchr any clues?

> I'm also now on the FBI watch list after searching "man kill".

:sweat_smile: :sweat_smile: :sweat_smile: 
Yeah, all of us are there. I'm writing so much text on killing workers. The FBI guys are like "what they're up to in that Parity" :laughing:

**Comment by mrcnski** [comment]
> My only concern is that the node will receive SIGKILL too, and what if some thread is writing to the database at the moment?

Good thinking, though any consensus-critical piece should already take this into account. I would be surprised if the DB wasn't properly ACID. The software can die at any time, whether for receiving a kill signal, plug being pulled, etc.

**Comment by s0me0ne-unkn0wn** [comment]
@bkchr okay, let's limit the scope of the task then. The idea was to try to shut down the whole node if a version mismatch was observed, killing the workers before that, so they could not produce an artifact after the node restart. There are two ways to achieve that: 1) to just `kill()` everything around and 2) to do a graceful shutdown propagating errors through all the pipelines up and down.

If we don't want to go hard, then maybe we shouldn't try to shut down? Reporting the situation at `error` loglevel and ceasing to accept any more worker jobs would do the trick. I've kind of convinced myself that stale workers cannot cause any harm, so probably it's the easiest solution?

**Comment by mrcnski** [comment]
> We should clearly not start sending any kill signals around :P If we detect a version mismatch, the most crucial thing to do is that the worker is not doing anything.

@bkchr In general we shouldn't be liberal with kill signals, but this seems to me like an exceptional scenario. It might actually be more likely for someone's node to die (for a number of reasons), with the same result as a kill, than for them to do an in-place upgrade. I would just kill the workers and not worry about it -- worst-case scenario we get a half-written temporary artifact. Or what other possibilities are you worried about? I definitely agree that we shouldn't complicate this -- increasing the maintenance burden of this code is likely to lead to more bugs and be a net negative outcome.

**Comment by bkchr** [comment]
> I would just kill the workers and not worry about it -- worst-case scenario we get a half-written temporary artifact.

I also said this or tried to say this. Killing the workers is fine, but they will kill themselves when they find the version mismatch? The parent process should only force kill them when it shuts down, but not try to find all processes that we may have started, even in earlier instances e.g `ps -a | grep polkadot`.




> Reporting the situation at `error` loglevel and ceasing to accept any more worker jobs would do the trick.

Should probably work for the beginning and then later we can improve this! 

> I've kind of convinced myself that stale workers cannot cause any harm, so probably it's the easiest solution?

If we have artifact ids with a random number as proposed by @koute it should be fine!

**Comment by s0me0ne-unkn0wn** [comment]
> If we have artifact ids with a random number as proposed by @koute it should be fine!

I modeled the scenarios that can take place and now I believe that even including a random number or a version number is redundant.

The preparation worker produces an artifact in a temporary file. It gets renamed to a real usable artifact on the node side. Only the node instance that started the worker can do such rename (because the worker signals to do the rename through a socket, and sockets are set up by the node instance on worker spawn).

Now, we have an old node that started an old worker. The worker has not concluded yet, and the node gets upgraded in place.

Any new worker started by the old node would report a version mismatch and exit immediately. It doesn't matter at this point if we kill the node or just spam errors into its logs. What matters is that at that point node has to be restarted, either by the node owner or by our code.

Now, if the old worker concludes before the node restart, an artifact is stored in the artifact cache. But the node cannot use it as it cannot spawn workers anymore because of version mismatch. When the node restart finally happens, the artifact cache is pruned, so the artifact cannot outlive the node upgrade.

The second scenario: the old worker does not conclude before the node restart. The new node is already started, and after that, the old worker produces a temporary file with the artifact. But that temporary file will never be renamed to a real usable artifact because that should be done by the node instance that is not there anymore. The only leftover is a useless temporary file which will be pruned either by the cache pruner or on the next node restart.

Let me know if I'm missing something but it seems that just ceasing processing new requests if the version mismatch condition is recognized is enough.

**Comment by s0me0ne-unkn0wn** [comment]
Pushed a very minimal solution: do not try shutting down, just log the situation and cease working. Should also check conditions of raising disputes: wouldn't workers exiting with error cause disputes?

File: node/core/pvf/src/execute/worker.rs
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/execute/worker.rs:268]: Also in tests, in several places scattered throughout the code. Each of them requires adding a build script to the crate,  so I decided just not to do version checking for tests to keep the footprint of changes minimal.

Comment (ordian) [reply node/core/pvf/src/execute/worker.rs:268]: Please add some docs :P

**Review by ordian — DISMISSED** [review]
LGTM

**Comment by ordian** [comment]
> Should also check conditions of raising disputes: wouldn't workers exiting with error cause disputes?

If I'm reading the code correctly, if the validation host is down, we return
https://github.com/paritytech/polkadot/blob/40ab88b583a2f836616df17b97e36c3000a5b88b/node/core/candidate-validation/src/lib.rs#L747
`InternalError` in candidate validation, which gets converted to an `Err` [here](https://github.com/paritytech/polkadot/blob/40ab88b583a2f836616df17b97e36c3000a5b88b/node/core/candidate-validation/src/lib.rs#L630)
In that case, approval-voting will result in [failed](https://github.com/paritytech/polkadot/blob/40ab88b583a2f836616df17b97e36c3000a5b88b/node/core/approval-voting/src/lib.rs#L2408) status, which doesn't raise a dispute. So this node will be a no-show for all candidates until a restart. It will also not issue dispute votes. Until restart.

**Review by eskimor — APPROVED** [review]

File: node/core/pvf/src/prepare/worker.rs
Comment (mrcnski) [inline node/core/pvf/src/prepare/worker.rs:331]: How can this mismatch occur? This should be thoroughly documented, maybe in its own e.g. `# Version Mismatch` section.

**Review by mrcnski — APPROVED** [review]
I still don't understand why we don't kill the node. Leaving it running still leaves open the possibility of disputes being raised, which is what we are trying to fix in the first place. Even if with the current architecture disputes wouldn't happen, and we test everything etc., something might change in the future breaking our assumptions. I would just kill the node for this exceptional case of an outdated node version... But I'll approve to not hold this up.

**Comment by s0me0ne-unkn0wn** [comment]
So, pushed a new minimal solution: check for version mismatch on the worker startup, and if the mismatch is detected, send `SIGKILL` to the parent process only.

Tests show that it works pretty fine: node is torn down, and when that happens, all its unix sockets get closed, which results in other workers getting an i/o error and exiting.

Before merging, I'd like to recruit a volunteer to test it under MacOS.

File: node/core/pvf/src/prepare/worker.rs
Comment (mrcnski) [inline node/core/pvf/src/prepare/worker.rs:369]: Can be extracted into a safe function (with the safety guarantee documented) -- to avoid use of `unsafe` here -- in `worker_common.rs`.

**Comment by mrcnski** [comment]
> Before merging, I'd like to recruit a volunteer to test it under MacOS.

I can do it. Have any specific testing instructions?

Also, is it possible to add some tests to the integration test suite in this module? I don't see why not. e.g. spawn a worker and send a version mismatch, spawn multiple valid workers and then spawn another one and send a version mismatch, etc.

NOTE: Right now `libc` is a Linux-only dependency, that needs to be fixed.

File: node/core/pvf/src/prepare/worker.rs
Comment (mrcnski) [reply node/core/pvf/src/prepare/worker.rs:369]: I like the simplicity of this approach BTW. No need to even send a response, just kill the parent from the child!

File: node/core/pvf/src/worker_common.rs
Comment (mrcnski) [inline node/core/pvf/src/worker_common.rs:404]: Awesome, great comment! Some non-blockers:

nit 1: Docstrings should start with `///` not `//`.
nit 2: I would call this `kill_parent_node_in_emergency` to make it clear that this only works if called from a child worker.

Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/worker_common.rs:404]: > Docstrings should start with `///` not `//`

But it's `pub(crate)`, does it make any sense to generate docs for it?

Comment (mrcnski) [reply node/core/pvf/src/worker_common.rs:404]: I often do `cargo doc --open --document-private-items` to help me read through a module. Maybe a less-niche usecase is viewing the function doc in an IDE. Also, a private function might one day become public... I just default to always using `///`, don't see why not.

**Comment by mrcnski** [comment]
I'm seeing unexpected behavior after completing step 8: `failed to spawn an execute worker: AcceptTimeout`. The workers are never spawned and the mismatch case is never triggered. Looking into it.

**Comment by mrcnski** [comment]
> I'm seeing unexpected behavior after completing step 8: failed to spawn an execute worker: AcceptTimeout. The workers are never spawned and the mismatch case is never triggered. Looking into it.

If I remove the node before replacing it, then the version mismatch detection
works correctly.

If I try to overwrite it with `cp -f` (which your instructions say not to do,
but operators may do it), then I get `AcceptTimeout`. My guess is that the new
process immediately exits, though we don't check for this in
`spawn_with_program_path`. This immediate exit seems to be known behavior on
MacOS, based on
https://stackoverflow.com/questions/67378106/mac-m1-cping-binary-over-another-results-in-crash.
It results in `SpawnErr`.

Also, I once randomly got this fun error: `Malformed Mach-o file`, which
also resulted in a `SpawnErr`. This `SpawnErr` is harmless -- it just causes us to
retry spawning the worker every 3s in an infinite loop.

My recommendations:

1. Catch process exiting (`handle` becoming invalid?) in `spawn_with_program_path` (can be a follow-up issue).
1. ~~Double check that `SpawnErr` does not result in disputes.~~ (checked above)
1. Add a test for immediately-dying workers if possible (`tests/it/worker_common.rs`).
2. I think documenting this in-code is not needed as this is a super edge case.

This seems like the best we can do. Then we should be good to go. These changes
should be quick to implement and not hold up this PR much longer. Actually, (1)
is probably not even needed. It just seems better to handle that case in a nice
way instead of letting it timeout -- can be a follow-up assigned to me.

**Comment by s0me0ne-unkn0wn** [comment]
@mrcnski, thanks a lot for testing and feedback! I'll have to leave (1) to you, as on Linux, `cp -f` results in an error, refusing to replace the file in use. Nevertheless, unlinking it and then copying a new one in its place works.

I believe `SpawnErr` is not a concern as it does not result in a dispute anyway. And upgrading a validator in place under MacOS is such a super-super-edge case... It would be a shame to introduce any new code just to handle that.

I'll try to add some integration tests later (I'll be ooo for several days).

**Comment by mrcnski** [comment]
> I'll try to add some integration tests later (I'll be ooo for several days).

wasmio right? Enjoy Barcelona!

**Comment by s0me0ne-unkn0wn** [comment]
bot merge

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-dispute-storm-the-postmortem/2550/1

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/ux-of-distributing-multiple-binaries-take-2/2854/3

**Pull request merged**
